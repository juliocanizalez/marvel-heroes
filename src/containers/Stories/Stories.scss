@use '../../styles/global';

$bg-title-image : 'https://static3.srcdn.com/wordpress/wp-content/uploads/2020/02/Dark-Avengers-Comic-Iron-Patriot.jpg';
.stories{
  @extend %marvel-font;
  text-align: center;
  &-title{
    h1{
      font-size: 2.6rem;
      margin: 2rem 0;
    }
  }
  .cards{
    &-content{
      display: grid;
      grid-template-columns: repeat(4, 1fr);
      gap: 1rem;
    }
  }
}
.error-message {
  color: #b12d2d;
}

.details{
  text-align: center;
  @extend %marvel-font;
}

.back-button {
  color: #fff;
  outline: none;
  cursor: pointer;
  background-color: global.$primary;
  height: 2rem;
  padding: 0.5rem 1rem;
  border: none;
  margin-right: 2rem;
}

.action-btn{
  color: #fff;
  outline: none;
  cursor: pointer;
  background-color: global.$primary;
  height: 2rem;
  padding: 0.5rem 1rem;
  border: none;
  margin-right: 1rem;
}

.image-header-div {
  color: white;
  height: 300px;
  display: flex;
  justify-content: center;
  align-items: flex-end;
  padding-bottom: 20px;

  background: -moz-linear-gradient(
      top,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;
  background: -webkit-gradient(
      linear,
      left top,
      left bottom,
      color-stop(0%, rgba(0, 0, 0, 0)),
      color-stop(59%, rgba(0, 0, 0, 0)),
      color-stop(100%, rgba(0, 0, 0, 0.65))
    ),
    url($bg-title-image) no-repeat;
  background: -webkit-linear-gradient(
      top,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;
  background: -o-linear-gradient(
      top,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;
  background: -ms-linear-gradient(
      top,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;
  background: linear-gradient(
      to bottom,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, 0) 59%,
      rgba(0, 0, 0, 0.65) 100%
    ),
    url($bg-title-image) no-repeat;

  background-position: top;
  background-size: cover;
}



.image-header-title {
  text-align: center;
}

.story-info {
  margin: 20px 15%;
  font-size: 22px;
  text-align: center;
}

.original-issue-title {
  text-align: center;
  font-size: 2.5rem;
  letter-spacing: 10px;
}

.comic-description {
  display: flex;
  flex-wrap: wrap;
  margin: 15px 15%;
}

.comic-image {
  margin: 10px 10px 0px;
  height: 350px;
  width: 200px;
  min-width: 250px;
  background-size: contain;
  background-position: center;
  background-repeat: no-repeat;
}

.comic-info {
  width: 65%;
  text-align: justify;
  margin-left: 5px;
  margin-top: 10px;
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  justify-content: center;
}

.comic-description-p {
  margin-bottom: 20px;
}

.comic-title {
  font-weight: bold;
  margin-bottom: 20px;
}

.comic-actions {
  display: flex;
}

.view-comic-button {
  margin: 10px;
  padding: 10px;
  color: white;
  text-align: center;
  text-decoration: none;
  font-size: 15px;
  width: 150px;
  cursor: pointer;
  border: none;
  background-color: global.$primary;
  box-shadow: 0 0 2px rgba(0, 0, 0, 0.16), 0 2px 4px rgba(0, 0, 0, 0.26);
  -webkit-transition: box-shadow 0.2s cubic-bezier(0.895, 0.03, 0.685, 0.22);
  transition: box-shadow 0.2s cubic-bezier(0.895, 0.03, 0.685, 0.22);
  text-transform: uppercase;
}

.view-comic-button:hover {
  box-shadow: 0 0 6px rgba(0, 0, 0, 0.16), 0 6px 12px rgba(0, 0, 0, 0.32);
  -webkit-transition: box-shadow 0.2s cubic-bezier(0.165, 0.84, 0.44, 1);
  transition: box-shadow 0.2s cubic-bezier(0.165, 0.84, 0.44, 1);
  background-color: #d30000;
  color: white;
}

.view-comic-button:focus {
  outline: 0;
}

.actions-title-div{
  margin: 1.4rem 0;
}

.related-items-div{
  margin-top: 1.4rem;
}