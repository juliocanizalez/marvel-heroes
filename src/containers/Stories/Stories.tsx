import React from 'react';

const Stories: React.FC = () => {
  return <div>stories</div>;
};

export default Stories;
