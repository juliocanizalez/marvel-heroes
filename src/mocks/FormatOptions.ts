interface IFormatOptions {
  name: string;
  value: string;
}
const FormatOptions = [
  {
    value: 'comic',
    name: 'Comic',
  },
  {
    value: 'magazine',
    name: 'Magazine',
  },
  {
    value: 'trade paperback',
    name: 'Trade Paperback',
  },
  {
    value: 'hardcover',
    name: 'Hardcover',
  },
  {
    value: 'digest',
    name: 'Digest',
  },
  {
    value: 'graphic novel',
    name: 'Graphic Novel',
  },
  {
    value: 'digital comic',
    name: 'Digital Comic',
  },
  {
    value: 'infinite comic',
    name: 'Infinite Comic',
  },
] as IFormatOptions[];

export default FormatOptions;
