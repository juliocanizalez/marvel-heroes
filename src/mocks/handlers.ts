/* eslint-disable import/extensions */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable arrow-body-style */
import { rest } from 'msw';
import {
  characterResponse,
  charactersResponse,
  comicResponse,
  comicsResponse,
  storiesResponse,
  storyResponse,
} from './testData';

const handlers = [
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/characters`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(charactersResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/characters/:idCharacter`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(characterResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/comics`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(comicsResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/comics/:idComic`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(comicResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/stories`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(storiesResponse()));
  }),
  rest.get(`${process.env.REACT_APP_API_URL}v1/public/stories/:idStory`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(storyResponse()));
  }),
  rest.get('*', (req, res, ctx) => {
    // eslint-disable-next-line no-console
    console.error(`Please add request handler for ${req.url.toString()}`);
    return res(ctx.status(500), ctx.json({ error: 'Please add handler' }));
  }),
];

export default handlers;
