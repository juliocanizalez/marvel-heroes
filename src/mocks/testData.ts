/* eslint-disable */
import ICharacter from '../interfaces/ICharacter';
import IComic from '../interfaces/IComic';
import IStory from '../interfaces/IStory';

const createGenericApiResponse = <T>(results: T[]) => ({
  data: {
    offset: 0,
    limit: 12,
    total: 1,
    count: 1,
    results,
  },
});

const emptyResponse = () => ({
  data: {
    offset: 0,
    limit: 12,
    total: 0,
    count: 0,
    results: [],
  },
});

const charactersResponse = () =>
  createGenericApiResponse<ICharacter>([
    {
      id: 1,
      name: 'spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [],
      comics: {} as any,
      stories: {} as any,
      events: {} as any,
      series: {} as any,
    },
  ]);

const characterResponse = () =>
  createGenericApiResponse<ICharacter>([
    {
      id: 1,
      name: 'spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [
        {
          type: 'detail',
          url: 'http://marvel.com/characters/2854/adam_warlock?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
        {
          type: 'wiki',
          url: 'http://marvel.com/universe/Warlock,_Adam?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
      ],
      comics: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Comic',
          },
        ],
      },
      stories: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Story',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
    },
  ]);

const comicsResponse = () =>
  createGenericApiResponse<IComic>([
    {
      id: 1,
      title: 'comic spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [],
      characters: {} as any,
      stories: {} as any,
      events: {} as any,
      series: {} as any,
      variants: {} as any,
      collections: {} as any,
      collectedIssues: {} as any,
      creators: {} as any,
      digitalId: 0,
      issueNumber: 0,
      variantDescription: '',
      isbn: '',
      upc: '',
      diamondCode: '',
      ean: '',
      issn: '',
      format: '',
      pageCount: 10,
      textObjects: [],
      dates: [],
      prices: [],
      images: [],
    },
  ]);

const comicResponse = () =>
  createGenericApiResponse<IComic>([
    {
      id: 1,
      title: 'comic spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [
        {
          type: 'detail',
          url: 'http://marvel.com/characters/2854/adam_warlock?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
        {
          type: 'wiki',
          url: 'http://marvel.com/universe/Warlock,_Adam?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
      ],
      characters: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Char Spiderman',
          },
        ],
      },
      stories: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Story',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
      variants: {} as any,
      collections: {} as any,
      collectedIssues: {} as any,
      creators: {} as any,
      digitalId: 0,
      issueNumber: 0,
      variantDescription: '',
      isbn: '',
      upc: '',
      diamondCode: '',
      ean: '',
      issn: '',
      format: '',
      pageCount: 10,
      textObjects: [],
      dates: [],
      prices: [],
      images: [],
    },
  ]);

const storiesResponse = () =>
  createGenericApiResponse<IStory>([
    {
      id: 1,
      title: 'story spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      comics: {} as any,
      characters: {} as any,
      events: {} as any,
      series: {} as any,
      creators: {} as any,
      originalIssue: {} as any,
      type: '',
    },
  ]);

const storyResponse = () =>
  createGenericApiResponse<IStory>([
    {
      id: 1,
      title: 'story spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      comics: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Comic',
          },
        ],
      },
      characters: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Char Spiderman',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
      creators: {} as any,
      originalIssue: {
        resourceURI: 'apiUrl/comics/1',
        name: 'Spiderman Comic',
      },
      type: '',
    },
  ]);

export {
  createGenericApiResponse,
  emptyResponse,
  charactersResponse,
  characterResponse,
  comicsResponse,
  comicResponse,
  storiesResponse,
  storyResponse,
};
