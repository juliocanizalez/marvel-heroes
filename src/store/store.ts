/* eslint-disable no-underscore-dangle */
import { createStore, combineReducers, StoreEnhancer } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import SearchReducer from '../reducers/search/SearchReducer';
import SearchComicReducer from '../reducers/searchComic/SearchComicReducer';
import ViewItemReducer from '../reducers/viewItem/ViewItemReducer';
import LocalItemsReducer from '../reducers/localItems/LocalItemsReducer';

type WindowWithDevTools = Window & {
  __REDUX_DEVTOOLS_EXTENSION__: () => StoreEnhancer<unknown, {}>;
};

const isReduxDevtoolsExtenstionExist = (
  arg: Window | WindowWithDevTools,
): arg is WindowWithDevTools => '__REDUX_DEVTOOLS_EXTENSION__' in arg;

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['localItems'],
};

const rootReducer = combineReducers({
  search: SearchReducer,
  searchComic: SearchComicReducer,
  viewItem: ViewItemReducer,
  localItems: LocalItemsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = createStore(
  persistedReducer,
  isReduxDevtoolsExtenstionExist(window) ? window.__REDUX_DEVTOOLS_EXTENSION__() : undefined,
);

export const persistor = persistStore(store);

export type IRootState = ReturnType<typeof rootReducer>;

export default store;
