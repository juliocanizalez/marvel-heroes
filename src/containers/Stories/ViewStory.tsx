import React, { useEffect, useRef, useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faChevronLeft,
  faEyeSlash,
  faBookmark as faBookmarkSolid,
  IconDefinition,
} from '@fortawesome/free-solid-svg-icons';
import { faBookmark as faBookmarkRegular } from '@fortawesome/free-regular-svg-icons';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import queryString from 'query-string';

import { IRootState } from '../../store/store';
import { get } from '../../api/FetchInfo';
import RouteNames from '../../routes/RoutePaths';
import { setViewItemAsyncContent } from '../../actions/ViewItem';
import IGenericApiResponse from '../../interfaces/IGenericApiResponse';
import IStory from '../../interfaces/IStory';
import IComic from '../../interfaces/IComic';
import CustomOrderedList from '../../components/CustomOrderedList/CustomOrderedList';
import Loading from '../../components/Loading/Loading';
import { addBookmark, hideLocalItem, removeBookmark } from '../../actions/LocalItems';
import './Stories.scss';
import { imagePlaceHolder } from '../../utils/utils';

const ViewStoryPage = () => {
  interface PathParams {
    idStory: string;
  }

  const isMounted = useRef(true);
  const history = useHistory();
  const dispatch = useDispatch();
  const { idStory } = useParams<PathParams>();
  const {
    loading,
    error,
    data: genericResponse,
  } = useSelector((state: IRootState) => state.viewItem);
  const { hiddenItems, bookmarks } = useSelector((state: IRootState) => state.localItems);

  const { data } = genericResponse ?? {};
  const { results } = data ?? {};
  const [story, setStory] = useState<IStory | null>(null);
  const [originalIssue, setOriginalIssue] = useState<IComic | null>(null);
  const [inBookmark, setInBookmark] = useState(
    !!bookmarks.find((item) => item.type === 'STORY' && item.id === parseFloat(idStory)),
  );
  const [bookmarkIcon, setBookmarkIcon] = useState<IconDefinition>(
    inBookmark ? faBookmarkSolid : faBookmarkRegular,
  );
  const [hidden] = useState(
    !!hiddenItems.find((item) => item.type === 'STORY' && item.id === parseFloat(idStory)),
  );

  const handleBack = () => {
    if (history.length <= 2) history.push(RouteNames.HOME);
    else history.goBack();
  };

  const handleViewChar = (id: string) => history.push(`/characters/${id}`);

  const handleViewComic = (id: string) => history.push(`/comics/${id}`);

  const handleAddBookmark = (id: number) => {
    dispatch(addBookmark({ id, type: 'STORY' }));
    setBookmarkIcon(faBookmarkSolid);
    setInBookmark(true);
  };

  const handleRemoveBookmark = (id: number) => {
    dispatch(removeBookmark({ id, type: 'STORY' }));
    setBookmarkIcon(faBookmarkRegular);
    setInBookmark(false);
  };

  const handleHideItem = (id: number) => {
    dispatch(hideLocalItem({ id, type: 'STORY' }));
    handleBack();
  };

  useEffect(() => {
    if (results) {
      if (!('stories' in results[0])) {
        setStory(results[0] as unknown as IStory);
        const originalIssueMainData = results[0].originalIssue;
        if (originalIssueMainData) {
          const pathVars = originalIssueMainData.resourceURI.split('/');
          const comicId = pathVars[pathVars.length - 1];
          const url = `${
            process.env.REACT_APP_API_URL
          }v1/public/comics/${comicId}?${queryString.stringify({
            apikey: process.env.REACT_APP_PUBLIC_KEY,
          })}`;
          get<IGenericApiResponse<IComic>>(url)
            .then((res) => {
              const { data: comicData } = res ?? {};
              const { results: comicResult } = comicData ?? {};
              if (isMounted.current) {
                setOriginalIssue(comicResult[0]);
              }
            })
            .catch(() => {
              if (isMounted.current) {
                setOriginalIssue(null);
              }
            });
        }
      }
    }
  }, [results]);

  useEffect(() => {
    dispatch(setViewItemAsyncContent(true, '', null));
    const url = `${
      process.env.REACT_APP_API_URL
    }v1/public/stories/${idStory}?${queryString.stringify({
      apikey: process.env.REACT_APP_PUBLIC_KEY,
    })}`;
    get<IGenericApiResponse<IStory>>(url)
      .then((res) => {
        dispatch(setViewItemAsyncContent(false, '', res));
      })
      .catch(() => {
        dispatch(setViewItemAsyncContent(false, 'Could not load story', null));
      });

    return () => {
      isMounted.current = false;
    };
  }, []);

  return (
    <div className='details'>
      {loading && <Loading />}
      {error && <h2 className='error-message'>Could not load story ðŸ˜“</h2>}
      {!loading && !error && story && (
        <>
          <div className='image-header-div'>
            <button type='button' className='back-button' onClick={handleBack}>
              <FontAwesomeIcon icon={faChevronLeft} />
              {'\u00A0'}
              Back
            </button>
            <h1 className='image-header-title'>Marvel story</h1>
          </div>
          <div className='story-info mb-5'>{story.title}</div>
          <div className='story-info mb-5'>{story.description}</div>
          {originalIssue && (
            <>
              <h3 className='original-issue-title fade-in'>Original Issue</h3>
              <div className='comic-description'>
                <div
                  className='comic-image'
                  style={{
                    backgroundImage: `url('${
                      originalIssue.thumbnail
                        ? `${originalIssue.thumbnail.path}/portrait_uncanny.${originalIssue.thumbnail.extension}`
                        : imagePlaceHolder
                    }')`,
                  }}
                />
                <div className='comic-info'>
                  <p className='comic-title'>{originalIssue.title}</p>
                  <p className='comic-description-p'>{originalIssue.description}</p>
                  <div className='comic-actions'>
                    <button
                      type='button'
                      className='view-comic-button'
                      onClick={() => handleViewComic(originalIssue.id.toString())}
                    >
                      View comic
                    </button>
                  </div>
                </div>
              </div>
            </>
          )}
          <div className='actions-title-div'>
            <p className='actions-title'>Actions:</p>
          </div>
          <div className='actions-menu'>
            <button
              type='button'
              className='action-btn add-bookmark'
              onClick={() => {
                if (inBookmark) handleRemoveBookmark(parseFloat(idStory));
                else handleAddBookmark(parseFloat(idStory));
              }}
            >
              <FontAwesomeIcon icon={bookmarkIcon} />
              {'\u00A0'}
              {inBookmark ? 'Remove bookmark' : 'Add bookmark'}
            </button>
            {!hidden && (
              <button
                type='button'
                className='action-btn hide-item'
                onClick={() => {
                  handleHideItem(parseFloat(idStory));
                }}
              >
                <FontAwesomeIcon icon={faEyeSlash} />
                {'\u00A0'}
                Hide item
              </button>
            )}
          </div>
          <div className='related-items-div fade-in'>
            <h3 className='related-items-title'>Characters</h3>
            {story.characters.returned > 0 ? (
              <CustomOrderedList items={story.characters.items} onClick={handleViewChar} />
            ) : (
              <p className='not-found'>No characters found ðŸ¤”</p>
            )}
          </div>
          <div className='related-items-div fade-in'>
            <h3 className='related-items-title'>Comics</h3>
            {story.comics.returned > 0 ? (
              <CustomOrderedList items={story.comics.items} onClick={handleViewComic} />
            ) : (
              <p className='not-found'>No comics found ðŸ¤”</p>
            )}
          </div>
        </>
      )}
    </div>
  );
};

export default ViewStoryPage;
