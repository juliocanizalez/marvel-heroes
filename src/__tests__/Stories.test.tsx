/* eslint-disable arrow-body-style */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-unused-vars */
/* eslint-disable import/extensions */
/* eslint-disable no-undef */
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { screen, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';

import store from '../store/store';
import server from '../mocks/server';
import { emptyResponse } from '../mocks/testData';
import Stories from '../containers/Stories/Stories';

describe('Test on ListStoriesPage', () => {
  test('Should render component correctly', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Stories />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
      expect(container).toMatchSnapshot();
    });
  });

  test('should add/remove bookmark', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Stories />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
    });

    userEvent.click(container.querySelector('.btn-bookmark') as Element);

    expect(container.querySelector('.bookmark-selected')).not.toBe(null);

    userEvent.click(container.querySelector('.btn-bookmark') as Element);

    expect(container.querySelector('.bookmark-selected')).toBe(null);
  });

  test('should hide card and show hidden stories message', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Stories />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
      expect(container.querySelector('.loading-container')).toBeNull();
    });

    userEvent.click(container.querySelector('.btn-hide') as Element);

    expect(container.querySelector('.cards')).toBeNull();
    expect(screen.getByText(/Stories on this page are hidden 🤐/i)).toBeInTheDocument();
  });

  test('should handle no results', async () => {
    server.use(
      rest.get(`${process.env.REACT_APP_API_URL}v1/public/stories`, (req, res, ctx) => {
        return res(ctx.status(200), ctx.json(emptyResponse()));
      }),
    );

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Stories />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(screen.getByText(/Stories not found 😮/i)).toBeInTheDocument();
      expect(container.querySelector('.cards')).toBeNull();
      expect(container.querySelector('.pagination-options')).toBeNull();
    });
  });

  test('should handle api error', async () => {
    server.use(
      rest.get(`${process.env.REACT_APP_API_URL}v1/public/stories`, (req, res, ctx) => {
        return res(ctx.status(404));
      }),
    );

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Stories />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(screen.getByText(/Could not load stories 😓/i)).toBeInTheDocument();
      expect(container.querySelector('.cards')).toBeNull();
      expect(container.querySelector('.pagination-options')).toBeNull();
    });
  });
});
