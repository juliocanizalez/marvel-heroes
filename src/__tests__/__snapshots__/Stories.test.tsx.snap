// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on ListStoriesPage Should render component correctly 1`] = `
<div>
  <div
    class="stories"
  >
    <div
      class="stories-title"
    >
      <h1>
        Stories
      </h1>
    </div>
    
    <div
      class="cards"
    >
      <div
        class="cards-content"
      />
    </div>
    <div
      class="pagination-options"
    >
      <div
        class="multi-button"
      />
    </div>
    <div
      class="current-pagination-page"
    >
      <p>
        <b>
          Page: 
        </b>
        1 / 9611
      </p>
    </div>
  </div>
</div>
`;
